Use leak sanitizer instead of internal mdebug to check for memory leaks

levitte opened the pull request:
The leak sanitizer gives better reports (complete stack traces) and
works as a wrapper around the application instead of relying on
cooperative enabling and disabling calls (which are too easy to get
unbalanced).

Related to #8322

levitte committed:
Use leak sanitizer instead of internal mdebug to check for memory leaks

The leak sanitizer gives better reports (complete stack traces) and
works as a wrapper around the application instead of relying on
cooperative enabling and disabling calls (which are too easy to get
unbalanced).

Related to #8322

levitte commented:
This also replaces #8323

mspncp commented:
fly-by-comment:  typo in commit message title (and GitHub title): 'sasnitizer'

richsalz commented:
Should https://github.com/openssl/tools/blob/master/review-tools/opensslbuild#L36 also be changed?

levitte committed:
fixup! Use leak sanitizer instead of internal mdebug to check for memory leaks

levitte commented:
> fly-by-comment: typo in commit message title (and GitHub title): 'sasnitizer'

Fixed

I also changed the way it's enabled, as sufficiently old clang doesn't understand `-fsanitize=leak`, but the address sanitizer has understood to enable the leak sanitizer internally for longer.

levitte commented:
> Should https://github.com/openssl/tools/blob/master/review-tools/opensslbuild#L36 also be changed?

I think it should.  The leak sanitizer gives far better output than our mdebug ever will.

mattcaswell commented:
Travis failure looks relevant:

````
==34177==AddressSanitizer: detect_leaks is not supported on this platform.
````

levitte commented:
Damn...

t-j-h commented:
Drop that platform out for this check.
It is possible to build a mingw cross compiler with asan support - but it isn't necessary to get the testing we need here for most linux-like platforms.

levitte commented:
> Drop that platform out for this check.

I *am* worried about corner cases for code paths that only happen on platforms....  on the other hand, since this approach is clang/gcc centric already, maybe I worry too much.

levitte commented:
I'll rework this to avoid mingw...  that will require some major surgery of the build matrix

richsalz commented:
Are you still intending to drop the internal memory-leak tooling?  I totally support that in favor of ASAN.  ASAN is what everyone is doing. :)

levitte commented:
BTW, I have *no* idea why mingw was mentioned at all here.  The Travis failure happens on MacOS X

levitte commented:
The MacOS problem has been complained about here: https://github.com/google/sanitizers/issues/1026
Apparently, it comes down to clang...

richsalz commented:
When you remove the mdebug stuff, please also remove all the fault-insertion stuff in crypto/mem.c (shouldfail(), md_trace_fd, etc)

mattcaswell commented:
I think we should just disable this on MacOS.

levitte committed:
fixup! Use leak sanitizer instead of internal mdebug to check for memory leaks

levitte commented:
I solved the MacOS case by not specifying `ASAN_OPTIONS="detect_leaks=1"` in .travis.yml.  Leak detection is on by default in modern enough ASAN, I hope that makes it all reliable enough.

mspncp commented:
Sorry, I couldn’t resist any longer 😉.

levitte commented:
That's all right. Thanks

levitte commented:
I'm pondering if I should add a test program that intentionally leaks memory, for demonstration purposes...

paulidale reviewed (approved):
I'm inclined to think yes to the deliberately leaking test program.
We'd quickly find out if leak detection ever stopped working.

richsalz commented:
we have aborttest and memleaktest

levitte committed:
test/memleaktest.c: Modify for use with address/leak sanitizer

Detects if leak sanitizing is on, and directs the exit code accordingly.

Note that this program is designed to fail when leaking, as that's
expected, so to make it easy for wrapper scripts, we also make it look
like it fails when sanitizing isn't on.

levitte commented:
I just modified `test/memleaktest.c` to work with the sanitizer.

levitte committed:
fixup! test/memleaktest.c: Modify for use with address/leak sanitizer

levitte committed:
fixup! test/memleaktest.c: Modify for use with address/leak sanitizer

levitte commented:
> Travis failure looks relevant

Yup

levitte commented:
The appveyor failure doesn't seem relevant here

levitte commented:
CIs are happy.  Is @openssl happy?

paulidale reviewed (approved):
I'm okay with this.

richsalz commented on test/recipes/90-test_memleak.t:
```diff
@@ -10,6 +10,10 @@
 
 setup("test_memleak");
+
+plan skip_all => "MacOS currently doesn't support leak sanitizer"
```
Is isn't? https://clang.llvm.org/docs/LeakSanitizer.html

richsalz replied:
I guess it depends on which version of clang/llvm you have :(

levitte replied:
Yup.

richsalz commented:
are you going to remove the current mdebug cruft?

levitte commented:
> are you going to remove the current mdebug cruft?

That's my intention, yes.

levitte commented:
... or... wanna take that on? I wouldn't mind. You do need to retain the public API, though...

richsalz commented:
I will make a PR to remove the existing mem-debug stuff.  I had wanted to remove the malloc-fail stuff as well, but it seems there's interest in this now.  (I forget which comment in which PR)

richsalz commented:
PR #10572

levitte commented:
Merged.

22c22369bc09988eb8c6fdf4f543581a78fd3eb1 Use leak sanitizer instead of internal mdebug to check for memory leaks
ea7a952c8aab33d0bb0a2bbd210305d722a5702b test/memleaktest.c: Modify for use with address/leak sanitizer

levitte commented:
@richsalz, you've got a small update to do with test/memleaktest.c in #10572 :wink:

richsalz commented:
yes, i rebased, took master for the conflicts, squashed, and pushed.  i'll note that over there too.

mattcaswell commented:
This doesn't actually seem to work in travis. The memleak test is failing in the extended tests.

mattcaswell commented:
I'm proposing to temporarily revert this in #10689.
